python/ray/tune/execution/tune_controller.py

```diff
@@ -1432,9 +1436,6 @@ def _schedule_trial_stop(self, trial: Trial, exception: Optional[Exception] = No
         self._set_trial_status(trial, Trial.ERROR if exception else Trial.TERMINATED)
         trial.set_location(_Location())
 
-        if exception:
-            trial.handle_error(exc=exception)
```

Comment:
Key change 1: The reason for moving this `trial.handle_error` is:
* `handle_error` is what increments the number of failures.
* Upon a trial failure, we used to check `trial.should_recover` BEFORE incrementing the new failure, so `trial.num_failures` is 1 less than it should be at that check. (Let's say `num_failures=2, max_failures=3` at this point.)
* `handle_error` would happen afterwards right here. (`num_failures=3` now.)
* The old `should_recover` condition made it impossible for us to try recovering on a preemption error. (Even though `handle_error` would noop for the preemption error, we're already at `num_failures==max_failures=3`, so `should_recover=False`.)
* It's more intuitive to have `num_failures` updated by the time of `trial.should_recover`, so now we just handle the error separately.

Reply:
Seems like we need to call `trial.handle_error` for all the other places that are currently calling `_schedule_trial_stop`?

Reply:
In general though I think this does move us in the right direction. In the long-term we should clean up the state machine in the tune controller, in it's current state it's not really clear where error handling is supposed to take place. 😵‍💫

Reply:
Yep, I call it in the other 2 other places.